onnxruntime/test/providers/cpu/controlflow/loop_test.cc

```diff
@@ -574,6 +574,99 @@ TEST(Loop, InfiniteLoopTermination) {
+    auto& graph = model.MainGraph();
+
+    std::vector<NodeArg*> inputs;
+    std::vector<NodeArg*> outputs;
+
+    /* Never change cond_in so loop is infinite
+            Inputs: iter_num, cond_in, loop carried state variables.
+
+         iter_num_in    cond_in     [outer_scope_0]
+           (unused)        |                |
+                       [Identity]      [Identity]
+                           |               |
+                        cond_out     loop_var_0_out
+    */
+
+    // graph inputs types.
+    TypeProto int64_scalar;
+    int64_scalar.mutable_tensor_type()->set_elem_type(TensorProto_DataType_INT64);
+    int64_scalar.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(1);
+
+    TypeProto bool_scalar;
+    bool_scalar.mutable_tensor_type()->set_elem_type(TensorProto_DataType_BOOL);
+    bool_scalar.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(1);
+
+    TypeProto float_tensor;
+    float_tensor.mutable_tensor_type()->set_elem_type(TensorProto_DataType_FLOAT);
+    float_tensor.mutable_tensor_type()->mutable_shape()->add_dim();
+
+    // graph inputs
+    auto& iter_num_in = graph.GetOrCreateNodeArg("iter_num_in", &int64_scalar);
+    auto& cond_in = graph.GetOrCreateNodeArg("cond_in", &bool_scalar);
+
+    // outer scope value. need type but not shape.
+    auto& outer_scope_0 = graph.GetOrCreateNodeArg("outer_scope_0", &float_tensor);
+
+    // add so that we don't end up with it being considered a graph input
+    graph.AddOuterScopeNodeArg("outer_scope_0");
+
+    // graph outputs
+    auto& cond_out = graph.GetOrCreateNodeArg("cond_out", &bool_scalar);
+    auto& loop_var_0_out = graph.GetOrCreateNodeArg("loop_var_0_out", &float_tensor);
+
+    // cond_in -> cond_out
+    {
+      inputs = {&cond_in};
+      outputs = {&cond_out};
+
+      graph.AddNode("cond_in_identity", "Identity", "Forward cond_in to cond_out", inputs, outputs);
+    }
+
+    // outer_scope_0 -> loop_var_0_out
+    {
+      inputs = {&outer_scope_0};
+      outputs = {&loop_var_0_out};
+
+      graph.AddNode("loop_var_out", "Identity", "Forward outer_scope_0 to loop_var_0_out", inputs, outputs);
+    }
+
+    graph.SetInputs({&iter_num_in, &cond_in, &outer_scope_0});
+    graph.SetOutputs({&cond_out, &loop_var_0_out});
+
+    auto status = graph.Resolve();
+    EXPECT_EQ(status, Status::OK());
+
+    return graph.ToGraphProto();
+  };
+
+  LoopOpTester test{{}, create_subgraph};
+
+  test.AddInput<int64_t>("M", {1}, {1});
+  test.AddInput<bool>("cond", {1}, {true});
+  // TODO - when the type is float16 (input/output - loop_var_0_final), extra Cast nodes will be added 
+  //        and input type changed by InsertCastTransformer for graph execution which causes type mismatch failure. 
+  // Need to investigate how InsertCastTransformer works in future. 
+  test.AddInput<double>("fake", {1}, {0.f});
+  test.AddInput<float>("outer_scope_0", {1}, {kOuterNodeAddValue});
```

Comment:
I would have expected 'fake' to be used in the subgraph as well as outer_scope_0. could just be a pass through with an identity node. right now it's kind of a strange setup where the loop input disappears. harder for someone to understand the test do to that. 

To do that, in the subgraph do something similar to how outer_scope_0 and loop_var_0_out are handled. 

Also please test overriding the type of outer_scope_0 so it's double in the main graph and float in the subgraph.